Add support for setting up Python through renv

milanmlft opened the pull request:
Hi sandpaper devs!

I have been experimenting with ways to set up Python and manage Python dependencies in sandpaper lessons through [`renv::use_python()`](https://rstudio.github.io/renv/articles/python.html).  This is mainly geared towards running Python code chunks in RMarkdown files.

To see this in action, I set up a toy sandpaper lesson here: https://github.com/milanmlft/toy-sandpaper-python
Note that I also had to update the carpentries GitHub Actions slightly to get this to work: https://github.com/milanmlft/actions. If relevant, I can create a PR for that as well.

I thought this might be useful for other sandpaper users and would love to get your feedback!

This would also address #61 (I think).

zkamvar commented:
@milanmlft, THANK YOU SO MUCH FOR THIS PR! 

To manage expectations: I (at the moment, the sole maintainer) am quite overencumbered right now and will not be able to review this until mid-June 😞


That being said, I can see this being a real boon to some of our python lessons! I'm really happy that you've written the tests, functionality, and created a working demo 🤩. 

One comment I can make right now WRT to the actions: in https://github.com/carpentries/actions/commit/3532acf1d112ea1235e249ad6a2257ad9ab07833, instead of doing this, you can set `sandpaper: milanmlft/sandpaper@use_python` in your lesson config.yaml

Additionally, I think this is something @sabaferdous12 might be interested in.

zkamvar commented:
/document

milanmlft commented:
Hi @zkamvar, thanks for the nice feedback! And no worries about the review. For my own purposes, I can continue with my own fork for now. If anything else pops up, I'll add it here as well.

One problem I currently still have is that `update_cache()` does not work nicely with the Python dependencies. When `requirements.txt` contains some Python packages, but those are not actually installed, then `update_cache()` will remove those packages from `requirements.txt`, because `renv::update()` does not update Python dependencies and so when `renv::snapshot()` is called, it doesn't detect those packages and thus removes them from the `requirements.txt`. `manage_deps()` used to do this as well, but there I found a workaround, implemented in cea48bf.

Another note is the bug reported in rstudio/renv#1217, for which I implemented a workaround in 1495f89.

Oh and coincidentally, I'm actually working together with @sabaferdous12 on a project, which is exactly why I implemented this 😅

zkamvar commented:
There are some test failures here, but that's because they need to be skipped on Windows by the fact that the earlier context skips them on Windows. 

I'm still working on the documentation for contributors, but see here about tests: https://github.com/carpentries/sandpaper/pull/455#issuecomment-1550256711

I believe these are skipped on Windows because of the idiosyncracies with using {renv} on the windows runner (though I know it works well on Windows)

milanmlft commented:
Alright, I'll adapt the tests accordingly. I'm still doing some work on the tests anyway because I'm not entirely happy with them yet. I'll push some more updates in the coming days!

zkamvar commented:
> Alright, I'll adapt the tests accordingly. I'm still doing some work on the tests anyway because I'm not entirely happy with them yet. I'll push some more updates in the coming days!

Awesome! 

Also FYI (and for @sabaferdous12), #465 modifies the `render_html()` function, which I _believe_ was modified in the L2D fork of sandpaper, so be aware of that potential conflict when updating that fork.

milanmlft commented:
Hi @zkamvar, sorry for bothering, but do you have a clearer idea yet of when you'll be able to have a look at this?
If another approach is more suitable, I'm all ears :) 
Thank you!

zkamvar commented:
@milanmlft Thank you for reminding me about this. I apologise for the huge delay on this. Unfortunately, organizational priorites have shifted significantly since the beginning of May and I am currently in the process of shoring up capacity (i.e. documenation over at https://carpentries.github.io/workbench-dev/ and training @carpentries/workbench-maintainers to take over maintenance when my position ends in December; see https://carpentries.org/blog/2023/06/lesson-infrastructure-updates/#new-challenges).

I do think enabling authors to use python in the lesson is valuable and I believe I can carve out some time in August to review this, but because of the situation, I cannot make any hard promises.

For the review, would you briefly describe the workflow for someone developing a lesson using python with this? How do authors add, remove, update, and pin python package versions?

milanmlft commented:
Hi @zkamvar, thanks for the reply! And no worries about the delay. I heard about your funding running out, so I completely understand it must be a hectic time for you right now.

I described the workflow briefly below. Do you want me to add this to a vignette somewhere as well? 

***

## Workflow

### Creating a new lesson with Python support

`create_lesson()` now has an `add_python` argument to set up Python in the package cache of the lesson:

```r
tmp <- tempfile()
sandpaper::create_lesson(tmp, name = "This Lesson", add_python = TRUE)
```

A specific Python version can be specified by passing the path through the `python` argument (default: `NULL`). If `python = NULL`, the user will be prompted to select an appropriate version. In non-interactive sessions, `renv` will look for a an appropriate Python version automatically (see [`renv::use_python()`](https://rstudio.github.io/renv/reference/use_python.html) for more details).

### Adding Python support to an existing lesson

To add Python to the package cache of an existing lesson, simply call

```r
sandpaper::use_python()
```

within an existing lesson. This will update the `renv.lock` file in the `lesson-requirements` profile to track the Python version used for this lesson. A specific Python version can again be specified through the `python` argument, as described above.

### Adding and updating Python packages

Activating Python will make `renv` aware of any Python packages *installed* in the lesson's environment. However, unlike with R packages, `renv` cannot detect Python packages that are *used* but not installed.

So to add new Python packages, we still need to install them manually first.  The `py_install()` helper takes care of this:

```r
sandpaper::py_install(packages = c("<PY_PKG1>", "<PY_PKG2>", ...))
```

This will install the requested packages in the correct environment and add them to a `requirements.txt` file, which is tracked by `renv`. Use `<package>==<version>"` to install a specific version of a package. Once the package is recorded in `requirements.txt`, the package cache will track it and make sure it's installed with `sandpaper::manage_deps()`. 

Alternatively, you can also manually add package names to the `requirements.txt` file and let `sandpaper::manage_deps()` take care of the installation.

zkamvar commented:
Hi @milanmlft! Apologies again for the delay, but I am happy to say that the @carpentries/workbench-maintainers and I have gone through this today and we will report back on our testing.

zkamvar commented:
NOTE: I have not yet reviewed the code yet because I know things are wonky with {renv}. I wanted to give you the result of my own testing: 

In my test, I was able to get things running as expected locally. I modified your example to create a new python lesson and then add two python blocks demonstrating numpy and plotnine (in the reprex). I've pushed it to https://github.com/zkamvar/test-python. The output of which can be found in [test-python-output.zip](https://github.com/carpentries/sandpaper/files/12562818/test-python-output.zip)

<details>
<summary>output of reprex</summary>

``` r
library("sandpaper")
out <- "test-python"
sandpaper::create_lesson(out, open = FALSE, add_python = TRUE)
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…
#> ℹ No schedule set, using Rmd files in 'episodes/' directory.
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…→ To remove this message, define your schedule in 'config.yaml' or use `set_episodes()` to generate it.
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…────────────────────────────────────────────────────────────────────────────────
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…ℹ To save this configuration, use
#> 
#> set_episodes(path = path, order = ep, write = TRUE)
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…✔ First episode created in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python/episodes/introduction.Rmd'
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…ℹ Workflows up-to-date!
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…ℹ Consent to use package cache provided
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…
#> - The project is out-of-sync -- use `renv::status()` for details.
#> → Searching for and installing available dependencies
#> The following packages were discovered:
#> 
#> # ~/R/library ----------------------------------------------------------------
#> - base64enc     0.1-3
#> - bslib         0.5.1
#> - cachem        1.0.8
#> - cli           3.6.1
#> - digest        0.6.33
#> - ellipsis      0.3.2
#> - evaluate      0.21
#> - fastmap       1.1.1
#> - fontawesome   0.5.2
#> - fs            1.6.3
#> - glue          1.6.2
#> - highr         0.10
#> - htmltools     0.5.6
#> - jquerylib     0.1.4
#> - jsonlite      1.8.7
#> - knitr         1.43
#> - lifecycle     1.0.3
#> - magrittr      2.0.3
#> - memoise       2.0.1
#> - mime          0.12
#> - R6            2.5.1
#> - rappdirs      0.3.3
#> - rlang         1.1.1
#> - rmarkdown     2.24
#> - sass          0.4.7
#> - stringi       1.7.12
#> - stringr       1.5.0
#> - tinytex       0.46
#> - vctrs         0.6.3
#> - xfun          0.40
#> - yaml          2.3.7
#> 
#> They will be linked into the project library.
#> 
#> - Linking packages into the project library ... Done!
#> - Hydrated 31 packages in 58 milliseconds.
#> - The project is out-of-sync -- use `renv::status()` for details.
#> → Recording changes in lockfile
#> The following package(s) will be updated in the lockfile:
#> 
#> # CRAN -----------------------------------------------------------------------
#> - bslib         [* -> 0.5.1]
#> - cachem        [* -> 1.0.8]
#> - digest        [* -> 0.6.33]
#> - evaluate      [* -> 0.21]
#> - fontawesome   [* -> 0.5.2]
#> - fs            [* -> 1.6.3]
#> - htmltools     [* -> 0.5.6]
#> - jquerylib     [* -> 0.1.4]
#> - knitr         [* -> 1.43]
#> - rlang         [* -> 1.1.1]
#> - rmarkdown     [* -> 2.24]
#> - sass          [* -> 0.4.7]
#> - tinytex       [* -> 0.46]
#> - vctrs         [* -> 0.6.3]
#> - xfun          [* -> 0.40]
#> - yaml          [* -> 2.3.7]
#> 
#> # RSPM -----------------------------------------------------------------------
#> - base64enc     [* -> 0.1-3]
#> - cli           [* -> 3.6.1]
#> - ellipsis      [* -> 0.3.2]
#> - fastmap       [* -> 1.1.1]
#> - glue          [* -> 1.6.2]
#> - highr         [* -> 0.10]
#> - jsonlite      [* -> 1.8.7]
#> - lifecycle     [* -> 1.0.3]
#> - magrittr      [* -> 2.0.3]
#> - memoise       [* -> 2.0.1]
#> - mime          [* -> 0.12]
#> - R6            [* -> 2.5.1]
#> - rappdirs      [* -> 0.3.3]
#> - renv          [* -> 1.0.2]
#> - stringi       [* -> 1.7.12]
#> - stringr       [* -> 1.5.0]
#> 
#> The version of R recorded in the lockfile will be updated:
#> - R             [* -> 4.3.1]
#> 
#> - Lockfile written to "~/Documents/Carpentries/Git/zkamvar/test-python/renv/profiles/lesson-requirements/renv.lock".
#> → Adding `reticulate` as a dependency
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…
#> The following package(s) will be installed:
#> - here       [1.0.1]
#> - png        [0.1-8]
#> - Rcpp       [1.0.11]
#> - RcppTOML   [0.2.2]
#> - reticulate [1.31]
#> - rprojroot  [2.0.3]
#> - withr      [2.5.0]
#> These packages will be installed into "~/Documents/Carpentries/Git/zkamvar/test-python/renv/profiles/lesson-requirements/renv/library/R-4.3/x86_64-pc-linux-gnu".
#> 
#> # Installing packages --------------------------------------------------------
#> - Installing Rcpp ...                           OK [linked from cache]
#> - Installing RcppTOML ...                       OK [linked from cache]
#> - Installing rprojroot ...                      OK [linked from cache]
#> - Installing here ...                           OK [linked from cache]
#> - Installing png ...                            OK [linked from cache]
#> - Installing withr ...                          OK [linked from cache]
#> - Installing reticulate ...                     OK [linked from cache]
#> Successfully installed 7 packages in 19 milliseconds.
#> - Creating virtual environment 'renv-python-3.8' ... Done!
#> - Updating Python packages ... Done!
#> - Lockfile written to "~/Documents/Carpentries/Git/zkamvar/test-python/renv/profiles/lesson-requirements/renv.lock".
#> - Activated Python 3.8.8 [virtualenv; ./renv/profiles/lesson-requirements/renv/python/virtualenvs/renv-python-3.8]
#> ✔ Lesson successfully created in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python'
#> → Creating Lesson in '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python…
#> /home/zhian/Documents/Carpentries/Git/zkamvar/test-python
```

```` r
l <- pegboard::Lesson$new(out, jekyll = FALSE)
# Write some python example blocks ---------------------------------------------
pycode <- r"{

## Python examples


### Numpy

Fun fact! Numpy rhymes with "lumpy"

```{python}
import numpy as np
a = np.arange(6)
a2 = a[np.newaxis, :]
a2.shape
```

### Plotnine

When R users begrudgingly use python

```{python}
from plotnine import ggplot, geom_point, aes, stat_smooth, facet_wrap
from plotnine.data import mtcars

(ggplot(mtcars, aes("wt", "mpg", color="factor(gear)"))
 + geom_point()
 + stat_smooth(method="lm")
 + facet_wrap("~gear"))
```

}"
l$episodes[[1]]$add_md(pycode, where = 0) # add this at the top of the document
l$episodes[[1]]$write(fs::path(out, "episodes"), format = "Rmd")

# Add python dependencies ------------------------------------------------------
sandpaper::py_install(path = out, c("numpy", "plotnine"))
````

    #> - The project is out-of-sync -- use `renv::status()` for details.
    #> Using virtual environment '/home/zhian/Documents/Carpentries/Git/zkamvar/test-python/renv/profiles/lesson-requirements/renv/python/virtualenvs/renv-python-3.8' ...
    #> + /home/zhian/Documents/Carpentries/Git/zkamvar/test-python/renv/profiles/lesson-requirements/renv/python/virtualenvs/renv-python-3.8/bin/python -m pip install --upgrade --no-user numpy plotnine
    #> → Updating the package cache
    #> The following package(s) will be updated in the lockfile:
    #> 
    #> # CRAN -----------------------------------------------------------------------
    #> - lattice      [* -> 0.21-8]
    #> - Matrix       [* -> 1.6-1]
    #> - png          [* -> 0.1-8]
    #> - Rcpp         [* -> 1.0.11]
    #> - RcppTOML     [* -> 0.2.2]
    #> - reticulate   [* -> 1.31]
    #> 
    #> # RSPM -----------------------------------------------------------------------
    #> - here         [* -> 1.0.1]
    #> - rprojroot    [* -> 2.0.3]
    #> - withr        [* -> 2.5.0]
    #> 
    #> - Lockfile written to "~/Documents/Carpentries/Git/zkamvar/test-python/renv/profiles/lesson-requirements/renv.lock".
    #> The following will be written to requirements.txt:
    #> - backports.zoneinfo==0.2.1
    #> - contourpy==1.1.0
    #> - cycler==0.11.0
    #> - fonttools==4.42.1
    #> - importlib-resources==6.0.1
    #> - kiwisolver==1.4.5
    #> - matplotlib==3.7.2
    #> - mizani==0.9.3
    #> - numpy==1.24.4
    #> - packaging==23.1
    #> - pandas==2.0.3
    #> - patsy==0.5.3
    #> - Pillow==10.0.0
    #> - plotnine==0.12.3
    #> - pyparsing==3.0.9
    #> - python-dateutil==2.8.2
    #> - pytz==2023.3.post1
    #> - scipy==1.10.1
    #> - six==1.16.0
    #> - statsmodels==0.14.0
    #> - tzdata==2023.3
    #> - zipp==3.16.2
    #> 
    #> - Wrote Python packages to "~/Documents/Carpentries/Git/zkamvar/test-python/requirements.txt".
    sandpaper::build_lesson(out)

<sup>Created on 2023-09-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
</details>

## Interactive vs Non-interactive setup

One of the weird things that I found was that there was a difference if I set up the environment interactively vs setting it up via script. If I did it interactively, then I would be sent into anaconda hell where it would hang forever resolving dependencies, but non-interactively it worked as expected (and yes, I know not to use anaconda, but I wanted to try this out with a real-world python setup). I believe much of this can be addressed by instructing users how to use python virtual environments and providing a template repository. 

## Disappearing libraries

The other _really_ weird thing that I encountered was that if I used `sandpaper::serve()`, the files would build correctly on the first pass, but then if I modified the file with python code in it, the rebuild of that file would suddenly not be able to use any of the libraries. It goes away if I reset the site, but it's just kinda weird and I suspect it's a renv bug and I'm not sure how to address it. 

## GitHub python conflict

I cloned your toy python example and found that the setup did not work because [python 3.8.10 was not built for Ubuntu 22.04](https://github.com/zkamvar/toy-sandpaper-python/actions/runs/6122488273/job/16618362673): 


```
Run py_version <- ""
- One or more packages recorded in the lockfile are not installed.
- Use `renv::status()` for more details.
Warning message:
Project requested Python version '3.8' but '3.10' is currently being used 
Run actions/setup-python@v4
Installed versions
  Version 3.8.10 was not found in the local cache
  Error: The version '3.8.10' with architecture 'x64' was not found for Ubuntu 22.04.
  The list of all available versions can be found here: https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json
```

I checked the URL and sure enough, the earliest version of 3.8 was 3.8.12:

```
curl https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json \
| jq '.[] | select(.version | match("3.8")).files | map(select(.platform=="linux" and .platform_version=="22.04")) | .[].download_url'

"https://github.com/actions/python-versions/releases/download/3.8.18-5997368067/python-3.8.18-linux-22.04-x64.tar.gz"
"https://github.com/actions/python-versions/releases/download/3.8.17-5199874912/python-3.8.17-linux-22.04-x64.tar.gz"
"https://github.com/actions/python-versions/releases/download/3.8.16-3888430809/python-3.8.16-linux-22.04-x64.tar.gz"
"https://github.com/actions/python-versions/releases/download/3.8.15-3601364564/python-3.8.15-linux-22.04-x64.tar.gz"
"https://github.com/actions/python-versions/releases/download/3.8.14-3008344966/python-3.8.14-linux-22.04-x64.tar.gz"
"https://github.com/actions/python-versions/releases/download/3.8.13-2717995909/python-3.8.13-linux-22.04-x64.tar.gz"
"https://github.com/actions/python-versions/releases/download/3.8.12-2268496296/python-3.8.12-linux-22.04-x64.tar.gz"
```

I think this can be solved by checking for the minor version during the GitHub install. It's not perfect, but it will work.

zkamvar commented:
and again, sorry for all the feet dragging on this. I'm finally getting caught up.

milanmlft commented:
Hi @zkamvar, thanks for the feedback!
And no worries about the delay 🙂 

## Interactive vs Non-interactive setup

Interesting. To be honest, I haven't really tested it with anaconda environments. I just kept the [options from renv](https://rstudio.github.io/renv/reference/use_python.html) to specify which Python environment to use. Would it make sense to just remove the option to use `"conda"` and hardcode `type = "virtualenv"`?

## Disappearing libraries

That is very weird indeed! I will try looking into it if I find some time, but I don't really have an idea of what it might be at the moment.
It could be related to https://github.com/carpentries/sandpaper/pull/448#issuecomment-1534324832 but I'm not sure.

## GitHub Python conflict

> I think this can be solved by checking for the minor version during the GitHub install. It's not perfect, but it will work.

I agree! Currently the [GHA tries installing the Python version recorded in the `renv.lock` file](https://github.com/milanmlft/actions/blob/8f45e6e32b063509649a2d6ebcdaa66b72bf7516/setup-lesson-deps/action.yaml#L86-L98), but I wasn't aware of the version limits for the Action runners. Note also that the GHA workflows used for this example lesson come from [my personal fork of the carpentries actions](https://github.com/milanmlft/actions). Would it make sense to merge these with the carpentries actions (once everything is working)?

milanmlft commented:
Hi @zkamvar, I made a few updates to my `use_python` branch, specifically to avoid some *renv* side effects when running functions that call `renv::activate()` or `renv::load()` interactively. The problem I faced was that invoking these functions would cause *renv* to open the lesson's RStudio project, even when the lessons were stored in a temporary directory. This was quite annoying while running package tests interactively.

I tried a few different approaches to fix this but the on that seemed to work best was to run the *renv* functions in a separate R process through `callr::r()`. To avoid some code duplication, I also added the `wirth_renv_factory()` operator, which prepends any given function with the *renv* profile loading mechanism. The resulting function is then passed on to `callr::r()`. This nicely isolates the functions meant to be run in the *renv* environment. I currently used this only for the `use_python` related functions, but it could potentially be helpful for other parts of *sandpaper*. If so, let me know and I'll happily imlement that as well (possibly in a separate PR).

Let me know if you think this makes sense! I have no problem reverting this if you think it's overkill.

Thanks a lot!

milanmlft commented:
Pinging the current `sandpaper` maintainers @froggleston and @ErinBecker. I know that adding new features is not a priority right now, but would still kindly ask if you could have a look at this :) 
I think it would benefit lesson developers who want to create lessons with Python, and we've actually been using a version of this successfully in a few Carpentries-style lessons:
- https://learntodiscover.github.io/Basic_Python/
- https://learntodiscover.github.io/ML_supervised/
- https://learntodiscover.github.io/ML_unsupervised/

If there's anything I can do to speed up the process, please let me know!
Thanks!
